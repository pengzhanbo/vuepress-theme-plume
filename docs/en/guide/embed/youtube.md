---
title: YouTube Video
icon: mdi:youtube
createTime: 2025/10/08 14:30:33
permalink: /en/guide/embed/video/youtube/
---

## Overview

The theme provides the capability to embed YouTube videos.

This feature is powered by [vuepress-plugin-md-power](../../config/plugins/markdown-power.md).

## Configuration

This feature is disabled by default. You need to enable it in the theme configuration.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      youtube: true, // [!code ++]
    },
  })
})
```

## Syntax

Simple syntax:

```md
@[youtube](id)
```

More options:

```md
@[youtube autoplay loop start="0" end="0" width="100%" height="400px" ratio="16:9"](id)
```

**Parameter Description:**

- id: Video ID
- autoplay: Whether to enable autoplay
- loop: Whether to enable loop playback
- start: Video playback start time in seconds, or in `mm:ss` or `hh:mm:ss` format
- end: Video playback end time in seconds, or in `mm:ss` or `hh:mm:ss` format
- width: Video width
- height: Video height
- ratio: Video aspect ratio, defaults to `16:9`

## Examples

### Widescreen Video

Input:

```md
@[youtube](0JJPfz5dg20)
```

Output:

@[youtube](0JJPfz5dg20)
