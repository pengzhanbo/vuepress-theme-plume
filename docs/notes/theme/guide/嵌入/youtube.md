---
title: Youtube 视频
icon: mdi:youtube
createTime: 2024/03/28 14:30:33
permalink: /guide/embed/video/youtube/
---

## 概述

主题提供了 嵌入 Youtube 视频 的功能。

该功能由 [vuepress-plugin-md-power](../../config/plugins/markdownPower.md) 提供支持。

## 配置

该功能默认不启用。你需要在主题配置中开启。

::: code-tabs
@tab .vuepress/config.ts

```ts
export default defineUserConfig({
  theme: plumeTheme({
    plugins: {
      markdownPower: {
        youtube: true,
      },
    }
  })
})
```

:::

## 语法

简单的语法：

```md
@[youtube](id)
```

更多选项：

```md
@[youtube autoplay loop start="0" end="0" width="100%" height="400px" ratio="16:9"](id)
```

**选项说明：**

- id: 视频 ID
- autoplay: 是否自动播放
- loop: 是否循环播放
- start: 视频开始播放时间点，单位为秒， 也可以使用 `mm:ss` 或 `hh:mm:ss` 格式
- end: 视频结束播放时间点，单位为秒， 也可以使用 `mm:ss` 或 `hh:mm:ss` 格式
- width: 视频宽度
- height: 视频高度
- ratio: 视频比例，默认 `16:9`

## 示例

### 宽频视频

输入：

```md
@[youtube](0JJPfz5dg20)
```

输出：

@[youtube](0JJPfz5dg20)
