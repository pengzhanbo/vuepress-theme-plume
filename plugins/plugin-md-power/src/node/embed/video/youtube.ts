/**
 * @[youtube](id)
 */
import type { PluginWithOptions } from 'markdown-it'
import type { YoutubeTokenMeta } from '../../../shared/index.js'
import { URLSearchParams } from 'node:url'
import { parseRect } from '../../utils/parseRect.js'
import { resolveAttrs } from '../../utils/resolveAttrs.js'
import { timeToSeconds } from '../../utils/timeToSeconds.js'
import { createEmbedRuleBlock } from '..//createEmbedRuleBlock.js'

const YOUTUBE_LINK = 'https://www.youtube.com/embed/'

export const youtubePlugin: PluginWithOptions<never> = (md) => {
  createEmbedRuleBlock<YoutubeTokenMeta>(md, {
    type: 'youtube',
    name: 'video_youtube',
    syntaxPattern: /^@\[youtube([^\]]*)\]\(([^)]*)\)/,
    meta([, info = '', id = '']) {
      const { attrs } = resolveAttrs(info)

      return {
        id,
        autoplay: attrs.autoplay ?? false,
        loop: attrs.loop ?? false,
        start: timeToSeconds(attrs.start),
        end: timeToSeconds(attrs.end),
        title: attrs.title,
        width: attrs.width ? parseRect(attrs.width) : '100%',
        height: attrs.height ? parseRect(attrs.height) : '',
        ratio: attrs.ratio ? parseRect(attrs.ratio) : '',
      }
    },
    content(meta) {
      const params = new URLSearchParams()

      if (meta.autoplay) {
        params.set('autoplay', '1')
      }

      if (meta.loop) {
        params.set('loop', '1')
      }

      if (meta.start) {
        params.set('start', meta.start.toString())
      }

      if (meta.end) {
        params.set('end', meta.end.toString())
      }

      const source = `${YOUTUBE_LINK}/${meta.id}?${params.toString()}`

      return `<VideoYoutube src="${source}" width="${meta.width}" height="${meta.height}" ratio="${meta.ratio}" title="${meta.title}" />`
    },
  })
}
