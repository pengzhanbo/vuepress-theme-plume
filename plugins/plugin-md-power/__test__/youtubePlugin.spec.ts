import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { youtubePlugin } from '../src/node/embed/video/youtube.js'

function createMarkdown() {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(youtubePlugin)
}

describe('youtubePlugin', () => {
  it('should work', () => {
    const md = createMarkdown()
    const code = `\
@[youtube](123456)

@[youtube loop autoplay title="test"](123456)

@[youtube autoplay start="40" end="1:20"](123456)

@[youtube width="100%" height="600px"](123456)

@[youtube width="100%" ratio="16:9"](123456)
`

    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()
    const code = `\
@[youtube]()

@[youtube]xxx

@[ youtube]123456

@[ youtube](123456)
`

    expect(md.render(code)).toMatchSnapshot()
  })
})
