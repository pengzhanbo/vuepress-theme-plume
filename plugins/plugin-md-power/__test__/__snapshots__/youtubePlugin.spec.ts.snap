// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`youtubePlugin > should not work 1`] = `
"<VideoYoutube src="https://www.youtube.com/embed//?" width="100%" height="" ratio="" title="undefined" /><p>@[youtube]xxx</p>
<p>@[ youtube]123456</p>
<p>@<a href="123456"> youtube</a></p>
"
`;

exports[`youtubePlugin > should work 1`] = `"<VideoYoutube src="https://www.youtube.com/embed//123456?" width="100%" height="" ratio="" title="undefined" /><VideoYoutube src="https://www.youtube.com/embed//123456?autoplay=1&loop=1" width="100%" height="" ratio="" title="test" /><VideoYoutube src="https://www.youtube.com/embed//123456?autoplay=1&start=40&end=80" width="100%" height="" ratio="" title="undefined" /><VideoYoutube src="https://www.youtube.com/embed//123456?" width="100%" height="600px" ratio="" title="undefined" /><VideoYoutube src="https://www.youtube.com/embed//123456?" width="100%" height="" ratio="16:9" title="undefined" />"`;
